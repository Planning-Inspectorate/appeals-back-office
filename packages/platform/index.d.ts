import { HandlerFunction } from 'got';

declare module '@pins/platform' {
	export function createTtlHandler(): HandlerFunction;
	export function validatePostcode(value: string): boolean;
	export function validatePastDate(value: Date | string | number): boolean;
	export function validateFutureDate(value: Date | string | number): boolean;
	export function yesterday(): Date;
}

export interface LocalPlanningDepartment {
	FID: number;
	LPA21CD: string;
	LPA21NM: string;
	Co_terminous: 'Y' | 'N';
}
