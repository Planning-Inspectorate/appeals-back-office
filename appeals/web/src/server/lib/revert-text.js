import { REVERT_BUTTON_TEXT } from '@pins/appeals/constants/common.js';

/**
 *
 * @param {string} finalCommentsType
 * @returns
 */
export const findButtonText = (finalCommentsType) => {
	switch (finalCommentsType) {
		case 'lpa':
			return REVERT_BUTTON_TEXT.LPA_FINAL_COMMENT;
		case 'appellant':
			return REVERT_BUTTON_TEXT.APPELLANT_FINAL_COMMENT;
		default:
			return REVERT_BUTTON_TEXT.DEFAULT_TEXT;
	}
};
