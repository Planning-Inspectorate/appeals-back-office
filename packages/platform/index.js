export * from './http/ttl.js';
export * from './util/date.js';
export * from './validators/date.js';
export * from './validators/postcode.js';
