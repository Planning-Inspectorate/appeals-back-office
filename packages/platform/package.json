{
  "name": "@pins/platform",
  "version": "0.0.0",
  "type": "module",
  "main": "index.js",
  "private": true,
  "engines": {
    "node": ">=22.0.0 <= 22.14.0"
  },
  "scripts": {
    "clean": "rimraf .turbo node_modules dist",
    "lint:js": "npx eslint .",
    "lint:js:fix": "npx eslint . --fix",
    "test": "cross-env NODE_ENV=test NODE_OPTIONS=\"--experimental-vm-modules --expose-gc\" npx jest --logHeapUsage --ci",
    "test:cov": "cross-env NODE_ENV=test NODE_OPTIONS=\"--experimental-vm-modules --expose-gc\" npx jest --logHeapUsage --ci --coverage",
    "tscheck": "npx tsc -p jsconfig.json --maxNodeModuleJsDepth 0"
  },
  "dependencies": {
    "date-fns": "2.28.0",
    "dotenv": "16.0.3",
    "got": "^12.5.2",
    "html": "^1.0.0",
    "lodash-es": "^4.17.21",
    "node-html-parser": "^5.3.3",
    "randexp": "^0.5.3"
  },
  "devDependencies": {
    "@types/html": "^1.0.4",
    "@types/lodash-es": "^4.17.12",
    "cross-env": "^7.0.3",
    "jest": "^29.7.0",
    "type-fest": "^4.37.0"
  }
}
