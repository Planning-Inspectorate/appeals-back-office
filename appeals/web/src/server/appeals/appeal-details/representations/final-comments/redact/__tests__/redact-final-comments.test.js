/* eslint-disable jest/expect-expect */
import { findButtonText } from '#lib/revert-text.js';
import {
	appealDataFullPlanning,
	finalCommentsForReview
} from '#testing/app/fixtures/referencedata.js';
import { createTestEnvironment } from '#testing/index.js';
import { parseHtml } from '@pins/platform';
import nock from 'nock';
import supertest from 'supertest';

const { app, installMockApi, teardown } = createTestEnvironment();
const request = supertest(app);
const baseUrl = '/appeals-service/appeal-details';
const finalCommentsTypes = [
	{
		type: 'appellant',
		label: 'appellant',
		origin: 'citizen'
	},
	{
		type: 'lpa',
		label: 'LPA',
		origin: 'lpa'
	}
];

describe('final-comments', () => {
	beforeEach(() => {
		installMockApi();

		nock('http://test/')
			.get('/appeals/2')
			.reply(200, {
				...appealDataFullPlanning,
				appealId: 2,
				appealStatus: 'final_comments'
			})
			.persist();
	});
	afterEach(teardown);

	describe('GET /', () => {
		for (const finalCommentsType of finalCommentsTypes) {
			it(`should render the redact ${finalCommentsType.type} final comments page with the expected content`, async () => {
				nock('http://test/')
					.get(`/appeals/2/reps?type=${finalCommentsType.type}_final_comment`)
					.reply(200, finalCommentsForReview)
					.persist();

				const response = await request.get(
					`${baseUrl}/2/final-comments/${finalCommentsType.type}/redact`
				);

				expect(response.statusCode).toBe(200);

				const element = parseHtml(response.text);

				expect(element.innerHTML).toMatchSnapshot();

				const unprettifiedHTML = parseHtml(response.text, {
					skipPrettyPrint: true,
					rootElement: 'body'
				}).innerHTML;

				expect(unprettifiedHTML).toContain(
					`<a href="/appeals-service/appeal-details/2/final-comments/${finalCommentsType.type}" class="govuk-back-link">Back</a>`
				);
				expect(unprettifiedHTML).toContain('Appeal 351062</span>');
				expect(unprettifiedHTML).toContain(`Redact ${finalCommentsType.label} final comments</h1>`);
				expect(unprettifiedHTML).toContain('Original final comments:</p>');
				expect(unprettifiedHTML).toContain(
					`id="original-comment" class="govuk-inset-text govuk-!-margin-top-2"><div class="pins-show-more" data-label="Read more" data-mode="html">Awaiting final comments review</div></div>`
				);
				expect(unprettifiedHTML).toContain('Awaiting final comments review</textarea>');
				expect(unprettifiedHTML).toContain('Redact selected text</button>');
				const buttonText = findButtonText(finalCommentsType.type);
				expect(unprettifiedHTML).toContain(`${buttonText}</button>`);
				expect(unprettifiedHTML).toContain('Undo changes</button>');
				expect(unprettifiedHTML).toContain('Continue</button>');
			});
		}
	});

	describe('POST /', () => {
		for (const finalCommentsType of finalCommentsTypes) {
			it(`should redirect to the redact ${finalCommentsType.type} final comments check your answers page`, async () => {
				nock('http://test/')
					.get(`/appeals/2/reps?type=${finalCommentsType.type}_final_comment`)
					.reply(200, finalCommentsForReview)
					.persist();

				const response = await request
					.post(`${baseUrl}/2/final-comments/${finalCommentsType.type}/redact`)
					.send({
						redactedRepresentation: 'Test redacted final comment text'
					});

				expect(response.statusCode).toBe(302);
				expect(response.text).toBe(
					`Found. Redirecting to /appeals-service/appeal-details/2/final-comments/${finalCommentsType.type}/redact/confirm`
				);
			});
		}
	});

	describe('GET /confirm', () => {
		for (const finalCommentsType of finalCommentsTypes) {
			it(`should render a 500 error page if redactedRepresentation is not present in the session (${finalCommentsType.type} final comments)`, async () => {
				nock('http://test/')
					.get(`/appeals/2/reps?type=${finalCommentsType.type}_final_comment`)
					.reply(200, finalCommentsForReview)
					.persist();

				const response = await request.get(
					`${baseUrl}/2/final-comments/${finalCommentsType.type}/redact/confirm`
				);

				expect(response.statusCode).toBe(500);

				const unprettifiedHTML = parseHtml(response.text, { skipPrettyPrint: true }).innerHTML;

				expect(unprettifiedHTML).toContain('Sorry, there is a problem with the service</h1>');
			});

			it(`should render the check your answers page with the expected content if redactedRepresentation is present in the session (${finalCommentsType.type} final comments)`, async () => {
				nock('http://test/')
					.get(`/appeals/2/reps?type=${finalCommentsType.type}_final_comment`)
					.reply(200, finalCommentsForReview)
					.persist();

				const redactedRepresentation = 'Test redacted final comment text';
				const redactPagePostResponse = await request
					.post(`${baseUrl}/2/final-comments/${finalCommentsType.type}/redact`)
					.send({
						redactedRepresentation
					});

				expect(redactPagePostResponse.statusCode).toBe(302);

				const response = await request.get(
					`${baseUrl}/2/final-comments/${finalCommentsType.type}/redact/confirm`
				);

				expect(response.statusCode).toBe(200);

				const unprettifiedHTML = parseHtml(response.text, { skipPrettyPrint: true }).innerHTML;

				expect(unprettifiedHTML).toContain(
					`Check details and accept ${finalCommentsType.label} final comments</h1>`
				);
				expect(unprettifiedHTML).toContain('Original final comments</dt>');
				expect(unprettifiedHTML).toContain(
					'class="pins-show-more" data-label="Read more" data-mode="html">Awaiting final comments review</div>'
				);
				expect(unprettifiedHTML).toContain('Redacted final comments</dt>');
				expect(unprettifiedHTML).toContain(
					`class="pins-show-more" data-label="Read more" data-mode="html">${redactedRepresentation}</div>`
				);
				expect(unprettifiedHTML).toContain(
					`href="/appeals-service/appeal-details/2/final-comments/${finalCommentsType.type}/redact?backUrl=/appeals-service/appeal-details/2/final-comments/${finalCommentsType.type}/redact/confirm">Change<span class="govuk-visually-hidden"> redacted final comments</span></a></dd>`
				);
				expect(unprettifiedHTML).toContain('Supporting documents</dt>');
				expect(unprettifiedHTML).toContain(`No documents</dd>`);
				expect(unprettifiedHTML).toContain('Review decision</dt>');
				expect(unprettifiedHTML).toContain('Redact and accept final comments</dd>');
				expect(unprettifiedHTML).toContain(
					`href="/appeals-service/appeal-details/2/final-comments/${finalCommentsType.type}?backUrl=/appeals-service/appeal-details/2/final-comments/${finalCommentsType.type}/redact/confirm">Change<span class="govuk-visually-hidden"> review decision</span></a></dd>`
				);
				expect(unprettifiedHTML).toContain(
					`Accept ${finalCommentsType.label} final comments</button></form>`
				);
			});
		}
	});

	describe('POST /confirm', () => {
		for (const finalCommentsType of finalCommentsTypes) {
			it(`should call the patch representation by ID endpoint with a status of 'valid' and redirect to the case details page (${finalCommentsType.type} final comments)`, async () => {
				nock('http://test/')
					.get(`/appeals/2/reps?type=${finalCommentsType.type}_final_comment`)
					.reply(200, finalCommentsForReview)
					.persist();

				const redactedRepresentation = 'Test redacted final comment text';
				const redactPagePostResponse = await request
					.post(`${baseUrl}/2/final-comments/${finalCommentsType.type}/redact`)
					.send({
						redactedRepresentation
					});

				expect(redactPagePostResponse.statusCode).toBe(302);

				const mockedPatchRepresentationEndpoint = nock('http://test/')
					.patch(`/appeals/2/reps/3670`)
					.reply(200, {
						...finalCommentsForReview.items[0],
						origin: finalCommentsType.origin,
						status: 'valid'
					});

				const response = await request
					.post(`${baseUrl}/2/final-comments/${finalCommentsType.type}/redact/confirm`)
					.send({});

				expect(mockedPatchRepresentationEndpoint.isDone()).toBe(true);
				expect(response.statusCode).toBe(302);
				expect(response.text).toBe(`Found. Redirecting to /appeals-service/appeal-details/2`);
			});
		}
	});
});
