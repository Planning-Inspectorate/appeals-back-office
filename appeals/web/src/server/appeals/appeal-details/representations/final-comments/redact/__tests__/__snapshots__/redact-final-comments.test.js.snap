// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`final-comments GET / should render the redact appellant final comments page with the expected content 1`] = `
"<main class="govuk-main-wrapper" id="main-content">
    <div class="govuk-grid-row">
        <div class="govuk-grid-column-two-thirds"><span class="govuk-caption-l">Appeal 351062</span>
            <h1 class="govuk-heading-l">Redact appellant final comments</h1>
        </div>
    </div>
    <div class="govuk-grid-row">
        <form method="POST" class="govuk-grid-column-two-thirds">
            <p class="govuk-body govuk-!-margin-bottom-0">Original final comments:</p>
            <div id="original-comment" class="govuk-inset-text govuk-!-margin-top-2">
                <div class="pins-show-more" data-label="Read more" data-mode="text">Awaiting final comments review</div>
            </div>
            <div class="govuk-form-group">
                <label class="govuk-label govuk-label--s" for="redact-textarea">Redacted final comments:</label>
                <textarea class="govuk-textarea" id="redact-textarea"
                name="redactedRepresentation" rows="5" readonly="true">Awaiting final comments review</textarea>
            </div>
            <div class="govuk-button-group">
                <button type="submit" class="govuk-button govuk-button--secondary" data-module="govuk-button"
                id="redact-button">Redact selected text</button>
                <button type="submit" class="govuk-button govuk-button--secondary"
                data-module="govuk-button" id="undo-button">Undo all changes</button>
            </div>
            <div class="govuk-button-group">
                <button type="submit" class="govuk-button" data-module="govuk-button">Continue</button>
            </div>
        </form>
    </div>
</main>"
`;

exports[`final-comments GET / should render the redact lpa final comments page with the expected content 1`] = `
"<main class="govuk-main-wrapper" id="main-content">
    <div class="govuk-grid-row">
        <div class="govuk-grid-column-two-thirds"><span class="govuk-caption-l">Appeal 351062</span>
            <h1 class="govuk-heading-l">Redact LPA final comments</h1>
        </div>
    </div>
    <div class="govuk-grid-row">
        <form method="POST" class="govuk-grid-column-two-thirds">
            <p class="govuk-body govuk-!-margin-bottom-0">Original final comments:</p>
            <div id="original-comment" class="govuk-inset-text govuk-!-margin-top-2">
                <div class="pins-show-more" data-label="Read more" data-mode="text">Awaiting final comments review</div>
            </div>
            <div class="govuk-form-group">
                <label class="govuk-label govuk-label--s" for="redact-textarea">Redacted final comments:</label>
                <textarea class="govuk-textarea" id="redact-textarea"
                name="redactedRepresentation" rows="5" readonly="true">Awaiting final comments review</textarea>
            </div>
            <div class="govuk-button-group">
                <button type="submit" class="govuk-button govuk-button--secondary" data-module="govuk-button"
                id="redact-button">Redact selected text</button>
                <button type="submit" class="govuk-button govuk-button--secondary"
                data-module="govuk-button" id="undo-button">Undo all changes</button>
            </div>
            <div class="govuk-button-group">
                <button type="submit" class="govuk-button" data-module="govuk-button">Continue</button>
            </div>
        </form>
    </div>
</main>"
`;
